perch/main.py

```diff
@@ -0,0 +1,103 @@
+import json
+import re
+import requests
+
+from .endpoints import ENDPOINTS
+
+
+class CrudBase(object):
+    def get(self):
+        raise NotImplementedError
+
+    def create(self):
+        raise NotImplementedError
+
+    def update(self):
+        raise NotImplementedError
+
+    def delete(self):
+        raise NotImplementedError
+
+
+class EndpointCrud(CrudBase):
+    def __init__(self, api, name=None, path=None, verbs=()):
+        self.api = api
+        self.name = name
+        self.path = path
+        for verb in verbs:
+            setattr(self, self.verb_to_name(verb), self.build_request(verb))
+
+    @staticmethod
+    def verb_to_name(verb):
+        verb = verb.lower()
+        name_map = {
+            'post': 'create',
+            'put': 'update'
+        }
+        return name_map.get(verb, verb)
+
+    def build_url(self, verb, **kwargs):
+        path = self.path
+        params = re.findall(':(.[a-z+_]*)/*', path)
```

Comment:
FYI: [`re.sub`](https://docs.python.org/3/library/re.html#re.sub) accepts a replacement function, so this could also be achieved with something like:
```py
def build_url(self, verb, **kwargs):
    def replace(match):
        param = match.group()
        arg = str(kwargs.get(param, ''))
        if not arg:
            if not path.endswith(param) or verb in ('put', 'delete',):
                raise TypeError('Squwaaaak! The {} kwarg is required!'.format(param))
        return param
    
    path = re.sub(':(.[a-z+_]*)/*', replace, self.path)
    url = self.api.root_url + path.rstrip('/')
    return url
```